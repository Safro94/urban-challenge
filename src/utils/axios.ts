import axios, { AxiosError, AxiosRequestConfig } from 'axios';
import https from 'https';

const opts: AxiosRequestConfig = {
	httpsAgent: new https.Agent({ rejectUnauthorized: false }),
	baseURL: process.env.REACT_APP_API_URL,
};

const interceptorError = (error: AxiosError) => Promise.reject(error.response);

const instance = axios.create(opts);

instance.interceptors.response.use(response => response, interceptorError);

export default instance;
