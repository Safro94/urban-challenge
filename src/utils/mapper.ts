import { IAvailableWorker, IAvailableWorkerResponse } from '../types';

export const mapAvailableWorkers = (
	workers: IAvailableWorkerResponse[]
): IAvailableWorker[] => {
	return workers.map((worker: IAvailableWorkerResponse) => ({
		slotId: worker.slot_id,
		availableWorkersIds: worker.availableWorker_ids,
	}));
};
