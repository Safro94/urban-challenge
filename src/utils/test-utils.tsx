import { FC, ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { BrowserRouter as Router } from 'react-router-dom';

import { theme } from '../theme';

const AllTheProviders: FC = ({ children }) => {
	return (
		<ThemeProvider theme={theme}>
			<Router>{children}</Router>
		</ThemeProvider>
	);
};

const customRender = (
	ui: ReactElement,
	options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';
export { customRender as render };
