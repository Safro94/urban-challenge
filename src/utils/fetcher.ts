import { AxiosResponse } from 'axios';
import {
	IAvailableWorkersResponse,
	ISlotsResponse,
	IWorkerResponse,
	RequestMethods,
} from '../types';

import axios from './axios';

interface IFetcher {
	method?: RequestMethods;
	url: string;
	data?: null | any;
	body?: null | any;
}

type Response = IWorkerResponse | ISlotsResponse | IAvailableWorkersResponse;

const fetcher = async ({
	method = RequestMethods.GET,
	url,
	body = null,
}: IFetcher): Promise<AxiosResponse<Response>> => {
	try {
		const res: AxiosResponse<Response> = await axios[method](url, body);
		return res;
	} catch (error) {
		throw error;
	}
};

export default fetcher;
