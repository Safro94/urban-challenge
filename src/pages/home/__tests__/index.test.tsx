import { render, waitFor } from '../../../utils/test-utils';

import Home from '..';

import WorkerService from '../../../services/workerService';

jest.mock('../../../containers/dateSelectionContainer', () => () => (
	<div>date selection container</div>
));

const mockSetWorkers = jest.fn();
jest.mock('../../../hooks/application', () => {
	return {
		useApplication: () => ({
			setWorkers: mockSetWorkers,
		}),
	};
});

jest.mock('../../../services/workerService', () => {
	return {
		getWorkers: jest.fn(),
	};
});

const mockHandleError = jest.fn();
jest.mock('react-error-boundary', () => ({
	...jest.requireActual('react-error-boundary'),
	useErrorHandler: () => mockHandleError,
}));

describe('Home', () => {
	const workers = [{ id: 1, name: 'test', rating: '4.5' }];

	beforeEach(() => {
		(WorkerService.getWorkers as jest.Mock).mockResolvedValue(workers);
	});

	it('should render correctly', async () => {
		const { container } = render(<Home />);

		await waitFor(() => {
			expect(container.firstChild).toMatchSnapshot();
		});
	});

	it('should call the workerService and setWorkers with the service response', async () => {
		render(<Home />);

		await waitFor(() => {
			expect(WorkerService.getWorkers).toHaveBeenCalledTimes(1);
			expect(WorkerService.getWorkers).toHaveBeenCalledWith(mockHandleError);

			expect(mockSetWorkers).toHaveBeenCalledTimes(1);
			expect(mockSetWorkers).toHaveBeenCalledWith(workers);
		});
	});
});
