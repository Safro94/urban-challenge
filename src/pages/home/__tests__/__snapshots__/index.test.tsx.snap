// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should render correctly 1`] = `
<div
  class="sc-bdfBQB gihfMK"
>
  <div
    class="sc-gsTEea XzNnr"
  >
    <div
      class="sc-dlfnuX kKuPIC"
    >
      <span
        class="sc-hKgJUU"
      >
        Matias Safranchik
      </span>
      <h2
        class="sc-eCstlR"
      >
        30 minutes appointment
      </h2>
      <p
        class="sc-jSgvzq"
      >
        A 30 minutes appointment with the best practitioners in the world. Select the date and time and enjoy the best experience of your life!
      </p>
    </div>
    <div>
      date selection container
    </div>
  </div>
</div>
`;
